src/probnum/kernels/_kernel.py

```diff
@@ -0,0 +1,302 @@
+"""Kernel / covariance function."""
+
+from typing import Callable, Generic, Optional, Tuple, TypeVar, Union
+
+import numpy as np
+import scipy.spatial
+
+import probnum.utils as _utils
+from probnum.type import IntArgType, ScalarArgType, ShapeArgType
+
+_InputType = TypeVar("InputType")
+
+
+class Kernel(Generic[_InputType]):
+    """Kernel / covariance function.
+
+    Kernels are a generalization of a positive-definite function or matrix. They
+    typically describe the covariance function of a random process and thus describe
+    its spatial or temporal variation. If evaluated at two sets of points a kernel
+    gives the covariance of the random process at these locations.
+
+    Parameters
+    ----------
+    input_dim :
+        Input dimension of the kernel.
+    output_dim :
+        Output dimension of the kernel.
+    kernelfun :
+        Function defining the kernel.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from probnum.kernels import Kernel
+    >>> # Data
+    >>> x = np.array([[1, 2], [-1, -1]])
+    >>> # Custom kernel from a (non-vectorized) covariance function
+    >>> k = Kernel(kernelfun=lambda x0, x1: (x0.T @ x1 - 1.0) ** 2, input_dim=2)
+    >>> k(x)
+    array([[16., 16.],
+           [16.,  1.]])
+    """
+
+    # pylint: disable="invalid-name"
+    def __init__(
+        self,
+        input_dim: IntArgType,
+        output_dim: IntArgType = 1,
+        kernelfun: Optional[
+            Callable[[_InputType, Optional[_InputType]], np.ndarray]
+        ] = None,
```

Comment:
what would you say is the preferred way of making a kernel? Initialising here, or subclassing? And how does a user know at this point, how the matrix decision is handled, i.e. what happens if the provided kernelfun is not vectorised?

Reply:
I decided to force the user to subclass Kernel by removing the initialization with `kernelfun` and making the class abstract.